workspaces/jenkins/plugins/scaffolder-backend-module-jenkins/src/config.ts

```diff
@@ -0,0 +1,128 @@
+import Jenkins from 'jenkins';
+
+const JENKINS_CONFIG_KEY: string = 'jenkins';
+const BASE_URL_KEY: string = 'baseUrl';
+const HEADERS_KEY: string = 'headers';
+const USERNAME: string = 'username';
+const API_KEY: string = 'apiKey';
+const CRUMB_ISSUER_ENABLED: string = 'crumbIssuerEnabled';
+
+const HTTP_PREFIX: string = 'http://';
+const HTTP_PROTOCOL: string = 'http://';
+const HTTPS_PROTOCOL: string = 'https://';
+
+export const jenkinsConfigStructure = {
+  baseUrl: 'string',
+  username: 'string',
+  apiKey: 'string',
+  crumbIssuerEnabled: 'boolean',
+  headers: 'Record<string, string> | null',
+};
+
+export class JenkinsConfig {
+  /**
+   * Must be something like http://hostname:port, https://hostname:port or https://hostname
+   */
+  baseUrl: string;
+
+  // Optional Keys
+  headers: Record<string, string> | null = null;
+  crumbIssuerEnabled: boolean | null = null;
+
+  constructor(
+    baseUrl: string,
+    headers: Record<string, string> | null,
+    crumbIssuerEnabled: boolean | null,
+  ) {
+    this.baseUrl = baseUrl;
+    this.headers = headers;
+    this.crumbIssuerEnabled = crumbIssuerEnabled;
+  }
+
+  static fromConfig(rootConfigService: RootConfigService): JenkinsConfig {
+    if (!rootConfigService.has(JENKINS_CONFIG_KEY)) {
+      throwError(BASE_URL_KEY, null);
+    }
+
+    const config = rootConfigService.getConfig(JENKINS_CONFIG_KEY);
+    if (
+      !config.has(BASE_URL_KEY) ||
+      !config.getString(BASE_URL_KEY).startsWith(HTTP_PREFIX)
+    ) {
+      throwError(BASE_URL_KEY, config.getString(BASE_URL_KEY));
+    }
+
+    let baseUrl = config.getString(BASE_URL_KEY);
+
+    if (config.has(USERNAME) && config.has(API_KEY)) {
+      if (baseUrl.startsWith(HTTP_PROTOCOL)) {
+        baseUrl = baseUrl.replace(
+          HTTP_PROTOCOL,
+          HTTP_PROTOCOL.concat(config.getString(USERNAME))
+            .concat(':')
+            .concat(config.getString(API_KEY))
+            .concat('@'),
+        );
+      } else {
+        baseUrl = baseUrl.replace(
+          HTTPS_PROTOCOL,
+          HTTPS_PROTOCOL.concat(config.getString(USERNAME))
+            .concat(':')
+            .concat(config.getString(API_KEY))
+            .concat('@'),
+        );
+      }
+    }
+
+    const headers =
+      config
+        .getOptionalConfig(HEADERS_KEY)
+        ?.keys()
+        .reduce((acc, key) => {
+          const value = config.getOptionalString(`${key}`);
+          if (value) {
+            acc[key] = value;
+          }
+          return acc;
+        }, {} as Record<string, string>) || null;
+
+    const crumbIssuerEnabled =
+      config.getOptionalBoolean(CRUMB_ISSUER_ENABLED) || true;
+
+    return new JenkinsConfig(baseUrl, headers, crumbIssuerEnabled);
+  }
+}
+
+export function buildJenkinsClient(config: JenkinsConfig) {
+  return new Jenkins({
+    baseUrl: config.baseUrl,
+    headers: config.headers,
+    crumbIssuer: config.crumbIssuerEnabled ? config.crumbIssuerEnabled : false,
+  });
+}
```

Comment:
```suggestion
export function buildJenkinsClient(config: RootConfigService) {
  const baseUrl = new URL(config.getString('jenkins.baseUrl'));
  url.username = config.getOptionalString('jenkins.username');
  url.password = config.getOptionalString('jenkins.apiKey');

  return new Jenkins({
    baseUrl: url.toString(),
    headers: config.getOptional<Record<string, string>>('jenkins.headers'),
    crumbIssuer: config.getOptionalBoolean('jenkins.crumbIssuerEnabled') ?? true,
  });
}
```

then all the rest in this file I think can go away

Reply:
This is better than actually solution, I will do It 🙏